lib/ansible/modules/crypto/openssl_privatekey.py

```diff
@@ -178,15 +178,28 @@ def generate(self, module):
 
             try:
                 privatekey_file = os.open(self.path,
-                                          os.O_WRONLY | os.O_CREAT | os.O_TRUNC,
-                                          self.mode)
+                                          os.O_WRONLY | os.O_CREAT | os.O_TRUNC)
 
                 if self.cipher and self.passphrase:
                     os.write(privatekey_file, crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey,
                                                                      self.cipher, to_bytes(self.passphrase)))
                 else:
                     os.write(privatekey_file, crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
                 os.close(privatekey_file)
+
+                if self.mode is None:
+                    self.mode = int('0600', 8)
+                else:
+                    try:
+                        self.mode = int(self.mode, 8)
+                    except ValueError as e:
+                        try:
+                            st = os.lstat(self.path)
+                            self.mode = AnsibleModule._symbolic_mode_to_octal(st, self.mode)
+                        except ValueError as e:
+                            module.fail_json(msg="%s" % to_native(e), exception=traceback.format_exc())
+
+                os.chmod(self.path, self.mode)
```

Comment:
I'd do this whole section like this:
``` python
# Assuming that self.mode was set in __init__()
privatekey_file = os.open(self.path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC)
os.close(privatekey_file)
if isinstance(self.mode, six.string_types):
    try:
        self.mode = int(self.mode, 8)
    except ValueError:
        try:
            st = os.lstat(self.path)
            self.mode = module._symbolic_mode_to_octal(st, self.mode)
        except ValueError as e:
            module.fail_json(msg="%s" % to_native(e), exception=traceback.format_exc())
os.chmod(self.path, self.mode)
privatekey_file = os.open(self.path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, self.mode)
if self.cipher and self.passphrase:
    os.write(privatekey_file, crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey,
                                                                                  self.cipher, to_bytes(self.passphrase)))
else:
    os.write(privatekey_file, crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
os.close(privatekey_file)
```

Just for my curiousity, I'm not sure why we're using os.open() instead of open().  Does os.open() make things more secure in this case?

Reply:
this requires two `os.open` calls, which I think is less efficient. I am not sure why `os.open` used instead of open, just came while refactoring.

Reply:
@MarkusTeufelberger @Spredzy What do you think about @abadger 's solution ?

Reply:
I wouldn't consider it the most elegant code, but it at least has no chance of exposing the private key with a wrong mode. I am not really concerned about in essence creating an empty file first. Performance wise, yes it'll be slower but I bet that the by far limiting factor of this module (next to network latency) is the private key generation, not a few filesystem calls. If this is actually of concern, the `_symbolic_mode_to_octal` function is something that might need a further look to work without a lstat call. For now I'd prefer @abadger's solution in combination with a default `0o600` mode - maaaaaybe even setting the initial mode already in `privatekey_file = os.open(self.path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC)` hardcoded to `0o600`.